core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableFilter.java

```diff
@@ -68,4 +74,15 @@ public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
     // EnumerableCalc is always better
     throw new UnsupportedOperationException();
   }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      RelTraitSet required) {
+    RelCollation collation = required.getTrait(RelCollationTraitDef.INSTANCE);
+    if (collation == null || collation.getFieldCollations().size() == 0) {
```

Comment:
should be assert?

Reply:
Ah maybe not.

There are several test cases that passed into null or empty fields collation. If not return null tests will fail.

Reply:
I remembered in `EnumerableConvention.enforce()` it has the same piece of code. :)

Reply:
Yes. I found similar code in `EnumerableConvention.enforce()`.

But, trait propagation in operators happens before `EnumerableConvention.enforce()`, thus EnumerableFilter and EnumerableProject will face first how to deal with this case. 

If 

1) assert. It means crash in operator, which is not ok.
2) if don't handle it, it leads NPE (when collation is null)

so seems best to me is return `null` when in such cases, which means cannot propagate traits (which also makes sense for null or empty collation).

And, I think  to have a really good solution/reasoning, it's better to first understand where does null or empty collation generated, and why. (which I don't have an  answer so far).

Reply:
Because some query or tests didn't add any trait defs. It is so bad decision to make Collation trait optional ❌. `ORDER BY` is sql standard, without supporting it, how can we call this a query optimizer? More freedom doesn't always make things better.